[WIP] Simpler IQL example

BY571 opened the pull request:
## Description

Rewrites the IQL online example and adds an offline training example for IQL.

## Motivation and Context

Updates the online IQL example similar to #967. Fixes the "correct_for_frame_skip" function for the nested config and adds a simple offline IQL example.

- [ ] I have raised an issue to propose this change ([required](https://github.com/pytorch/rl/issues) for new features and bug fixes)

## Types of changes

What types of changes does your code introduce? Remove all that do not apply:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds core functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)
- [X] Example (update in the folder of examples)

## Checklist

Go over all the following points, and put an `x` in all the boxes that apply.
If you are unsure about any of these, don't hesitate to ask. We are here to help!

- [X] I have read the [CONTRIBUTION](https://github.com/pytorch/rl/blob/main/CONTRIBUTING.md) guide (**required**)
- [ ] My change requires a change to the documentation.
- [ ] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [ ] I have updated the documentation accordingly.

vmoens commented on examples/iql/iql_online.py:
```diff
@@ -2,291 +2,138 @@
-
-            loss_td = loss_module(sampled_tensordict)
+        if collected_frames >= init_random_frames:
+            (actor_losses, q_losses, value_losses) = ([], [], [])
```
One way to do this is
```python
td_log = TensorDict({}, [num_updates])
for i in range(num_updates):
    loss_td = loss_module(sampled_tensordict)
    td_log[i] = loss_td.detach()
    ....
td_log = td_log.apply(lambda x: x.float().mean(), batch_size=[])

```
and then you can just log each value in that tensordict in wandb

vmoens reviewed (approved):
LGTM! Thanks for this!
Can you merge main into this?
I will try to launch them once you do!

vmoens commented:
@BY571 I forgot: Can you add the scripts in the tests too?

BY571 commented:
Updated online and added offline example tests for IQL and also added them for CQL as they were still missing. 
Also had to do some fixes here and there.

vmoens reviewed (approved):
LGTM

BY571 commented:
Updated the losses of IQL and CQL so that components can get called independently + small fixes here and there.
Online training results:
![image](https://github.com/pytorch/rl/assets/29492081/50dc6dae-f1f1-44be-9d87-3aac8af7b331)

Also doing some offline runs to check if performance fits with the paper for CQL.

Just saw that some examples failed in the tests when logging metrics - will check about that.

BY571 commented:
CQL offline version achieves now a better performance compared to the paper for hopper-medium (0.58 norm score):
![image](https://github.com/pytorch/rl/assets/29492081/c25d3d32-6662-4e28-b2f4-8ac352cd179a)

Implementation-wise wise should be fine now, lets see what the tests say.

vmoens commented on examples/cql/cql_offline.py:
```diff
@@ -37,49 +42,95 @@ def main(cfg: "DictConfig"):  # noqa: F821
+        # compute loss
+        loss_vals = loss_module(data.clone().to(device))
+
+        if i >= policy_eval_start:
+            actor_loss = loss_vals["loss_actor"]
+        else:
+            actor_loss = loss_vals["loss_actor_bc"]
```
this seems pretty custom, can we give a small comment on why we need it?
Also, wondering how that kind of control flow would fit within a trainer class

BY571 replied:
Indeed it's pretty custom to the CQL implementation. Just added it for completeness as it's in the official implementation. I'll add a comment on why it's there.

vmoens commented on examples/cql/offline_config.yaml:
```diff
@@ -1,37 +1,36 @@
   task: ""
   library: gym
-  exp_name: cql_halfcheetah-medium-v2
+  exp_name: cql_${replay_buffer.dataset}
```
Wow! Is this working?

vmoens commented on examples/iql/iql_online.py:
```diff
@@ -2,361 +2,188 @@
+            log_loss_td = TensorDict({}, [num_updates])
+            for j in range(num_updates):
+                # sample from replay buffer
+                sampled_tensordict = replay_buffer.sample().clone()
+                if sampled_tensordict.device != device:
+                    sampled_tensordict = sampled_tensordict.to(
+                        device, non_blocking=True
+                    )
+                else:
+                    sampled_tensordict = sampled_tensordict.clone()
```
there are more than one clone, do we need them all?

vmoens commented on examples/iql/utils.py:
```diff
@@ -0,0 +1,290 @@
+
+
+def make_offline_replay_buffer(rb_cfg):
+    data = D4RLExperienceReplay(
```
in our examples, we could default to the version that does not require the d4rl library, wdyt?
It's pretty annoying to install and the dataset works without it.

vmoens commented on test/test_cost.py:
```diff
@@ -4854,7 +4854,7 @@ def _create_seq_mock_data_cql(
 
     @pytest.mark.parametrize("delay_actor", (True, False))
-    @pytest.mark.parametrize("delay_qvalue", (True, False))
+    @pytest.mark.parametrize("delay_qvalue", (True, True))
```
typo :)

vmoens commented on torchrl/objectives/cql.py:
```diff
@@ -225,14 +222,33 @@ class _AcceptedKeys:
+        priority: NestedKey = "td_error"
+        cql_q1_loss: NestedKey = "cql_q1_loss"
+        cql_q2_loss: NestedKey = "cql_q2_loss"
         priority: NestedKey = "td_error"
```
duplicate

vmoens commented on torchrl/objectives/cql.py:
```diff
@@ -493,36 +501,62 @@ def forward(self, tensordict: TensorDictBase) -> TensorDictBase:
+        loss_actor, actor_metadata = self.actor_loss(td_device)
+        loss_alpha, alpha_metadata = self.alpha_loss(td_device)
+        metadata.update(
+            {
+                **bc_metadata,
+                **cql_metadata,
+                **actor_metadata,
+                **alpha_metadata,
+            }
```
this assumes that all metadata are Dict[Str, Any] but maybe we'll add smth else?
Why not 4 updates? (at least this version will crash if 2 metadata collide)

vmoens commented on torchrl/objectives/cql.py:
```diff
@@ -493,36 +501,62 @@ def forward(self, tensordict: TensorDictBase) -> TensorDictBase:
+            }
+        )
+        tensordict_reshape.set(
+            self.tensor_keys.priority, metadata.pop("td_error").detach().max(0)[0]
```
when using max(0) we should use `max(0).values` instead of `[0]`

vmoens commented on torchrl/objectives/cql.py:
```diff
@@ -741,54 +805,57 @@ def _loss_qvalue_v(self, tensordict: TensorDictBase) -> Tuple[Tensor, Tensor]:
+            )
 
-        td_error = abs(q_pred - target_value)
+        alpha_prime = torch.clamp(self.log_alpha_prime.exp(), min=0.0, max=1000000.0)
```
is `1000000.0` a  magic number? What is the purpose of this clamp (to understand how we can code this in a more grounded way)?
No need to clamp an exp to 0 below, we can just do clamp_max

BY571 replied:
Indeed the number seems quite random, I took it directly from the official implementation [1](https://github.com/aviralkumar2907/CQL/blob/d67dbe9cf5d2b96e3b462b6146f249b3d6569796/d4rl/rlkit/torch/sac/cql.py#L271).
As far as I could see it in some tests the alpha_prime value can become quite big I guess this is just to make sure its not exploding. But I'll have a look at the paper if they mention something more specific.

vmoens commented on torchrl/objectives/iql.py:
```diff
@@ -357,35 +361,37 @@ def forward(self, tensordict: TensorDictBase) -> TensorDictBase:
             "loss_qvalue": loss_qvalue.mean(),
             "loss_value": loss_value.mean(),
-            "entropy": -td_device.get(self.tensor_keys.log_prob).mean().detach(),
+            "entropy": -tensordict_reshape.get(self.tensor_keys.log_prob)
+            .mean()
+            .detach(),
```
I'd rather pre-compute that before putting it in the dict to avoid overloaded dict construction

vmoens reviewed (approved):
Wonderful, thanks a mil!
Just a couple of housekeeping comments

vmoens commented on examples/iql/iql_online.py:
```diff
@@ -117,7 +117,7 @@ def main(cfg: "DictConfig"):  # noqa: F821
                     )
                 else:
-                    sampled_tensordict = sampled_tensordict.clone()
+                    sampled_tensordict = sampled_tensordict
```
this can be removed :)

vmoens reviewed (approved):
Good one minor thing to edit and we're good I think

vmoens reviewed (approved):
Awesome let's proceed!
